ambry-cloud/src/main/java/com/github/ambry/cloud/VcrReplicationManager.java

```diff
@@ -202,23 +205,88 @@ protected void createThread(ReplicaThread replicaThread, boolean unused) {
    * Assign all replicas of a partition to the same thread to avoid concurrent updates to Azure
-   * @param remoteReplicaInfos List of {@link RemoteReplicaInfo}
-   * @param startThread start threads on creation
+   *
+   * @param partitionId Partition
+   * @param remoteReplicaInfos List of {@link RemoteReplicaInfo} List of replicas, must belong to same partition
    */
-  @Override
-  protected void addRemoteReplicaInfoToReplicaThread(List<RemoteReplicaInfo> remoteReplicaInfos, boolean startThread) {
+  protected void addPartitionToReplicaThread(PartitionId partitionId, List<RemoteReplicaInfo> remoteReplicaInfos)
+      throws ReplicationException {
+    /**
+     * This fn must be called with threadMapLock.writeLock() held.
+     * It is called by helix when a new partition (not a new replica) is assigned to VCR node.
+     * addPartition -> addPartitionToReplicaThread
+     */
+    long pid = partitionId.getId();
+    if (partitionIdThreadMap.containsKey(pid)) {
+      throw new ReplicationException(String.format("Partition-%s already exists in replica-thread-map", pid),
+          ServerErrorCode.IO_Error);
+    }
+    // All replicas belong to the same partition, so just assign all to the same thread
+    ReplicaThread rthread = threadPool.get(threadIndex.incrementAndGet() % threadPool.size());
     for (RemoteReplicaInfo rinfo : remoteReplicaInfos) {
       String datacenter = rinfo.getReplicaId().getDataNodeId().getDatacenterName();
       if (!localDatacenterName.equalsIgnoreCase(datacenter)) {
         throw new UnsupportedOperationException(
             String.format("Cross-colo replication from %s to %s is unsupported in VCR for replica {}",
                 localDatacenterName, datacenter, rinfo));
       }
-      ReplicaThread rthread = threadPool.get(threadIndex.incrementAndGet() % threadPool.size());
       rthread.addRemoteReplicaInfo(rinfo);
       rinfo.setReplicaThread(rthread);
-      rthread.startThread();
-      logger.info("Added replica {} to thread {}", rinfo, rthread.getName());
+      logger.info("[PARTITION] Added replica {} to thread {}", rinfo, rthread.getName());
+    }
+    partitionIdThreadMap.put(partitionId.getId(), rthread);
+    rthread.startThread();
+  }
+
+  /**
+   * Assign all replicas of a partition to the same thread to avoid concurrent updates to Azure
+   * @param remoteReplicaInfos List of {@link RemoteReplicaInfo} These replicas belong to different partitions
+   * @param unused unused variable
+   */
+  @Override
+  protected void addRemoteReplicaInfoToReplicaThread(List<RemoteReplicaInfo> remoteReplicaInfos, boolean unused) {
+    /**
+     * This fn must be called with threadMapLock.readLock() held.
+     * It is called by helix when a new replica (not a new partition) is assigned to VCR node.
+     * onReplicaAddedOrRemoved -> addPartitionToReplicaThread
+     *
+     * Cannot use the same rwlock.writeLock because this is called with rwlock.readLock.
+     * Lock upgrades are not allowed.
```

Comment:
you can change rwlock.readLock to rwLock.writeLock in ReplicationEngine. It's weird that modifying states only requires a read lock. It should be write lock anyway.

Reply:
i agree, but as the comment in onReplicaAddedOrRemoved indicates, it was done for performance reasons. although i don't think it affects performance drastically since the lock is held for a very short duration.